table/column.go

```diff
@@ -177,10 +177,46 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 		if err1 != nil {
 			logutil.BgLogger().Warn("Datum ToString failed", zap.Stringer("Datum", val), zap.Error(err1))
 		}
-		err = sc.HandleTruncate(types.ErrTruncatedWrongVal.GenWithStackByArgs(col.FieldType.CompactStr(), str))
-	} else {
-		err = sc.HandleTruncate(err)
+		err = types.ErrTruncatedWrongVal.GenWithStackByArgs(col.FieldType.CompactStr(), str)
+	} else if (sc.InInsertStmt || sc.InUpdateStmt) && !casted.IsNull() &&
+		(col.Tp == mysql.TypeDate || col.Tp == mysql.TypeDatetime || col.Tp == mysql.TypeTimestamp) {
+		tm := casted.GetMysqlTime()
+		mode := ctx.GetSessionVars().SQLMode
+
+		// in Mysql8.0+, the timestamp's case is special:
+		// no matter HasNoZeroDateMode or HasNoZeroInDateMode is enabled or not,
+		// if timestamp is invalid, an warning message whould be appended or error handling.
+		if (val.Kind() != types.KindMysqlTime || !val.GetMysqlTime().IsZero()) &&
+			((tm.IsZero() && mode.HasNoZeroDateMode()) || (tm.InvalidZero() && (mode.HasNoZeroInDateMode() || col.Tp == mysql.TypeTimestamp))) {
+			var (
+				zeroV types.Time
+				zeroT string
+			)
+			switch col.Tp {
+			case mysql.TypeDate:
+				zeroV, zeroT = types.ZeroDate, types.DateStr
+			case mysql.TypeDatetime:
+				zeroV, zeroT = types.ZeroDatetime, types.DateTimeStr
+			case mysql.TypeTimestamp:
+				zeroV, zeroT = types.ZeroTimestamp, types.TimestampStr
+			}
+
+			innerErr := types.ErrWrongValue.GenWithStackByArgs(zeroT, val.GetString())
+			ignoreErr := sc.DupKeyAsWarning
```

Comment:
Why use DupKeyAsWarning for ignore error? I think here it is not about dup key, but zero date?

Reply:
Because the value of `sc.DupKeyAsWarning` is the original value of ignoreError: https://github.com/pingcap/tidb/blob/43a5bb47977fc7952a8aaedffe0154e6918bf4ad/executor/executor.go#L1601-L1615